### PR TITLE
Log and handle shard state retrieval errors

### PR DESCRIPTION
**Pull request by harmony-ek**
## Issue

This is a partial WIP for #675.  Among other things, this PR eliminates the use of nil pointers returned upon failure.

### CODE REVIEW

File: core/resharding.go [hunk 1]
```diff
@@ -132,22 +133,47 @@ func GetEpochFromBlockNumber(blockNumber uint64) uint64 {
...
+		return nil, ctxerror.New("cannot get shard state for epoch",
+			"epoch", epoch).WithCause(err)
+	}
 
 	rndSeedBytes := bc.GetRandSeedByNumber(number)
+	if rndSeedBytes == [32]byte{} {
+		utils.GetLogInstance().Warn("cannot get randomness from block",
+			"blockNumber", number)
+		// TODO ek – should we fail here?  Continuing here would end up using
+		//  all-zeros for randomness, which seems like a security breach.
```
Comment (harmony-ek) [inline core/resharding.go]: @rlan35 what do you think?

File: core/resharding.go [hunk 2]
```diff
@@ -132,22 +133,47 @@ func GetEpochFromBlockNumber(blockNumber uint64) uint64 {
...
 	}
-	ss := GetShardingStateFromBlockChain(bc, epoch-1)
+	ss, err := GetShardingStateFromBlockChain(bc, epoch-1)
+	if err != nil {
+		utils.GetLogInstance().Crit("CANNOT CALCULATE SHARDING STATE",
+			"epoch", epoch)
+		// TODO ek – what should we do here?  We are processing a new epoch
+		//  block, so we better have the new sharding state before advancing
+		//  to the next block (which would require the new sharding state),
+		//  but the sharding state is not available here.
```
Comment (harmony-ek) [inline core/resharding.go]: @chaosma @rlan35, what do you think?

File: core/resharding.go [hunk 3]
```diff
@@ -132,22 +133,47 @@ func GetEpochFromBlockNumber(blockNumber uint64) uint64 {
 }
 
 // GetShardingStateFromBlockChain will retrieve random seed and shard map from beacon chain for given a epoch
-func GetShardingStateFromBlockChain(bc *BlockChain, epoch uint64) *ShardingState {
+func GetShardingStateFromBlockChain(bc *BlockChain, epoch uint64,
+) (*ShardingState, error) {
 	number := GetBlockNumberFromEpoch(epoch)
-	shardState := bc.GetShardStateByNumber(number)
+	shardState, err := bc.GetShardStateByNumber(number)
```
Comment (harmony-ek) [inline core/resharding.go]: BTW, we have both `ShardingState` and `ShardState` (confusing naming).  The former is a more comprehensive collection of sharding information including randomness, and the latter is just typedef-ed as a list of `Committee`-s.  Maybe we should rename `ShardState` to `Committees` to better reflect this.  Cc: @chaosma

File: core/resharding.go (hunk 2 above)
Comment (harmony-ek) [reply core/resharding.go]: BTW this question is essential to resolving this issue.

File: core/resharding.go (hunk 1 above)
Comment (rlan35) [reply core/resharding.go]: sure. sounds good. Leave it as TODO for now.

File: core/resharding.go (hunk 2 above)
Comment (rlan35) [reply core/resharding.go]: Leave it as todo for now. The resharding code needs some final tweaking. We still need more work to finalize the code.

**Pull request merged**